### PR TITLE
2018/11/07/01

### PR DESCRIPTION
**Pull request by arek1-618**
Unread messages and improving history.

Fix problems with a profile picture.

### CODE REVIEW

File: app/Chat.php [hunk 1]
```diff
@@ -556,69 +556,104 @@ public function getHistoryByType(string $roomType = 'global', ?int $messageId =
...
+		}
+		$query = (new Db\Query())
+			->select($param)
+			->from(['M' => static::TABLE_NAME['message'][$roomType]])
+			->leftJoin(
+				['R' => static::TABLE_NAME['room'][$roomType]],
+				'R.' . static::COLUMN_NAME['room'][$roomType] . ' = M.' . static::COLUMN_NAME['message'][$roomType] . " AND R.userid = {$userId}"
+			)
+			->where(['or', ['R.last_message' => null], ['<', 'R.last_message', new \yii\db\Expression('M.id')]]);
+		switch ($roomType) {
```
Comment (rskrzypczak) [inline app/Chat.php]: Twice?
switch ($roomType) {

File: public_html/layouts/basic/modules/Chat/resources/Chat.js [hunk 2]
```diff
@@ -1097,6 +1103,15 @@ window.Chat_JS = class Chat_Js {
 		});
 	}
 
+	/**
+	 * Register button unread.
+	 */
+	registerButtonUnread() {
+		this.container.find('.js-btn-unread').off('click').on('click', (e) => {
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Chat/resources/Chat.js]: .off('click') is required

File: public_html/layouts/basic/modules/Chat/resources/Chat.js [hunk 3]
```diff
@@ -407,21 +417,11 @@ window.Chat_JS = class Chat_Js {
...
-					view: 'Entries',
-					mode: 'history',
-					groupHistory: element.data('group-name')
-				}, false).done((html) => {
-					self.messageContainer.html(html);
-				});
-			})
-		})
+		this.container.find('.js-chat-nav-history .js-chat-link').each((index, element) => {
+			$(element).off('click').on('click', () => {
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Chat/resources/Chat.js]: .off('click')

File: layouts/basic/modules/Chat/Modal.tpl [hunk 4]
```diff
@@ -54,11 +68,11 @@
 					<span class="fas fa-users mr-2"></span>
 					{\App\Language::translate('LBL_ROOM_GROUP', $MODULE_NAME)}
 				</div>
-				<ul class="js-room-type u-font-size-13px p-0" data-room-type="group"
+				<ul class="js-room-type u-font-size-13px p-0" data-room-type="group" data-favorite="false"
 					data-js="data">
 					{foreach item=GROUP_NAME key=GROUP_ID from=\App\Fields\Owner::getInstance('CustomView')->getGroups(false)}
 						{assign var=TRANSLATE_GROUP value=\App\Language::translate($GROUP_NAME)}
-						{assign var=SELECTED value=$CURRENT_ROOM['recordId'] == $GROUP && $CURRENT_ROOM['roomType'] === 'group'}
+						{assign var=SELECTED value=isset($GROUP) && $CURRENT_ROOM['recordId'] == $GROUP && $CURRENT_ROOM['roomType'] === 'group'}
```
Comment (rskrzypczak) [inline layouts/basic/modules/Chat/Modal.tpl]: $GROUP what is it?

Comment (mariuszkrzaczkowski) [reply layouts/basic/modules/Chat/Modal.tpl]: GROUP_ID ?

File: layouts/basic/modules/Chat/Item.tpl [hunk 5]
```diff
@@ -3,17 +3,16 @@
...
-		{assign var=IMAGE value=$ROW['image']}
-		{assign var=IS_IMAGE value=isset($IMAGE['url'])}
+		{assign var=IS_IMAGE value=!empty($ROW['image'])}
 		<div class="u-w-50px {if $USER_ID == $ROW['userid']} mr-3 {else} ml-3 {/if} js-author text-center"
 			 data-user-name="{$ROW['user_name']}"
 			 data-role-name="{$ROW['role_name']}" data-js="data">
-			<div class="o-chat__img-container mx-auto {if !$IS_IMAGE} p-1 {/if}">
-				<img src="{if $IS_IMAGE}{$IMAGE['url']}{/if}" class="{if !$IS_IMAGE} hide{/if} o-chat__author-img"
+			<div class="js-image o-chat__img-container mx-auto {if !$IS_IMAGE} p-1 {/if}" data-js="copy">
+				<img src="{$ROW['image']}" class="{if !$IS_IMAGE} hide{/if} js-src o-chat__author-img"
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Chat/Item.tpl]: js-chat-image_src

File: layouts/basic/modules/Chat/Item.tpl [hunk 6]
```diff
@@ -3,17 +3,16 @@
...
 					 alt="{$ROW['user_name']}"
-					 title="{$ROW['user_name']}"/>
-				<span class="fas fa-user u-font-size-38px {if $IS_IMAGE} hide{/if} o-chat__author-name"
-					  title="{$ROW['user_name']}"></span>
+					 title="{$ROW['user_name']}" data-js="hide"/>
+				<span class="js-icon fas fa-user u-font-size-38px {if $IS_IMAGE} hide{/if} o-chat__author-name"
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Chat/Item.tpl]: js-chat-image_icon

File: layouts/basic/modules/Chat/Modal.tpl (hunk 4 above)
Comment (mariuszkrzaczkowski) [reply layouts/basic/modules/Chat/Modal.tpl]: ???

File: modules/Chat/views/Entries.php [hunk 7]
```diff
@@ -183,17 +184,26 @@ public function getForRecord(\App\Request $request)
...
 		$viewer->assign('CURRENT_ROOM', \App\Chat::getCurrentRoom());
-		$viewer->assign('HISTORY_GROUP', $chat->getHistoryByType($request->getByType('groupHistory', 2)));
 		$viewer->assign('CHAT_ENTRIES', $chatEntries);
 		$viewer->assign('SHOW_MORE_BUTTON', count($chatEntries) > \AppConfig::module('Chat', 'CHAT_ROWS_LIMIT'));
-		$viewer->view('History.tpl', $request->getModule());
+		$viewer->view('Entries.tpl', $request->getModule());
+	}
+
+	/**
+	 * @param \App\Request $request
```
Comment (mariuszkrzaczkowski) [inline modules/Chat/views/Entries.php]: ?

File: app/Chat.php [hunk 8]
```diff
@@ -556,69 +556,165 @@ public function getHistoryByType(string $roomType = 'global', ?int $messageId =
...
+				$query->select(['M.*', 'name' => 'RN.groupname', 'R.last_message', 'recordid' => "M.{$columnMessage}"])
+					->innerJoin(
+						['R' => static::TABLE_NAME['room'][$roomType]],
+						"R.{$columnRoom} = M.{$columnMessage} AND R.userid = {$userId}"
+					)
+					->leftJoin(['RN' => 'vtiger_groups'], "RN.groupid = M.{$columnMessage}");
+				break;
+			case 'global':
+				$query->select(['M.*', 'name' => 'RN.name', 'R.last_message', 'recordid' => "M.{$columnMessage}"])
+					->leftJoin(
```
Comment (rskrzypczak) [inline app/Chat.php]: 3 times the same join:
> ->innerJoin(
> 	['R' => static::TABLE_NAME['room'][$roomType]],
> 	"R.{$columnRoom} = M.{$columnMessage} AND R.userid = {$userId}"
> )

move before the switch

File: app/Chat.php [hunk 9]
```diff
@@ -556,69 +556,165 @@ public function getHistoryByType(string $roomType = 'global', ?int $messageId =
...
+						['R' => static::TABLE_NAME['room'][$roomType]],
+						"R.{$columnRoom} = M.{$columnMessage} AND R.userid = {$userId}"
+					)
+					->leftJoin(['RN' => 'u_#__chat_global'], "RN.global_room_id = M.{$columnMessage}");
+				break;
+		}
+		return $query->where(['or', ['R.last_message' => null], ['<', 'R.last_message', new \yii\db\Expression('M.id')]])
```
Comment (rskrzypczak) [inline app/Chat.php]: why 'M.id' is in \yii\db\Expression class?

File: layouts/basic/modules/Chat/Modal.tpl [hunk 10]
```diff
@@ -1,120 +1,108 @@
...
-									</a>
-								</div>
-							</li>
-
+				{foreach item=GROUP_ROOM key=KEY from=\App\Chat::getRoomsByUser()}
+					{assign var=LBL_GROUP_ROOM value="LBL_ROOM_$KEY"|upper}
+					<div class="text-uppercase bg-color-grey-200 p-2 font-weight-bold js-group-name" data-js="data"
+						 data-group="{$KEY}">
+						{if $KEY === 'crm'}<span class="fas fa-star mr-2"></span>{/if}
+						{if $KEY === 'group'}<span class="fas fa-users mr-2"></span>{/if}
```
Comment (rskrzypczak) [inline layouts/basic/modules/Chat/Modal.tpl]: {elseif ..}

File: tests/Base/Z_Chat.php [hunk 11]
```diff
@@ -98,6 +94,55 @@ private static function getKeyMessage($entries, int $id)
 		return false;
 	}
 
+	/**
+	 * Create User.
+	 *
+	 * @codeCoverageIgnore
+	 *
+	 * @param $login
```
Comment (mariuszkrzaczkowski) [inline tests/Base/Z_Chat.php]: ?

File: tests/Base/Z_Chat.php [hunk 12]
```diff
@@ -98,6 +94,55 @@ private static function getKeyMessage($entries, int $id)
...
+	 *
+	 * @return mixed
+	 */
+	private static function createUser($login)
```
Comment (mariuszkrzaczkowski) [inline tests/Base/Z_Chat.php]: unnecessary >> A_User::createUsersRecord()

File: tests/Base/Z_Chat.php [hunk 13]
```diff
@@ -98,6 +94,55 @@ private static function getKeyMessage($entries, int $id)
...
+	 * Setting of tests.
+	 */
+	public static function setUpBeforeClass()
+	{
+		static::$chatActive = \App\Module::isModuleActive('Chat');
+		if (!static::$chatActive) {
+			(new \Settings_ModuleManager_Module_Model())->enableModule('Chat');
+		}
+		\App\User::setCurrentUserId(\App\User::getActiveAdminId());
+		$recordModel = \Vtiger_Record_Model::getCleanInstance('Contacts');
```
Comment (mariuszkrzaczkowski) [inline tests/Base/Z_Chat.php]: unnecessary >>   C_RecordActions::createContactRecord()

**Pull request merged**